feat: add redaction for sensitive variables in completion suggestions

fixing-things-enjoyer opened the pull request:
## Description

This pull request introduces a new feature to prevent sensitive environment variable values from being displayed in completion suggestions.

A new global environment variable, fish_redact_vars, is introduced. It can be set to a list of glob patterns (e.g., *_TOKEN, *_SECRET*). Any variable name matching one of these patterns will have its value replaced with [redacted] in the completion description. This helps to avoid accidental exposure of secrets like API keys or tokens when using tab completion in a public setting.

For example:
`set -g fish_redact_vars '*_TOKEN' '*_SECRET*'`

## Screenshots
(I'm pressing the Tab key at the end of `MY_`)

This:
<img width="605" height="203" alt="image" src="https://github.com/user-attachments/assets/9dd2a673-de6b-493e-b1a5-022ffad5ab2d" />


Becomes this:
<img width="1043" height="283" alt="image" src="https://github.com/user-attachments/assets/3b9ec009-68e5-4dbf-816b-f55985da4593" />





This change also includes updates to the localization files to make the new `Variable: [redacted]` string available for translation by the community.

Fixes issue # (N/A)

## TODOs:
<!-- Just check off what what we know been done so far. We can help you with this stuff. -->
- [x] Changes to fish usage are reflected in user documentation/manpages.
- [x] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.rst <!-- Don't document changes for completions inside CHANGELOG.rst, there are lot of such edits -->

krobelus commented on doc_src/interactive.rst:
```diff
@@ -56,6 +56,15 @@
+
+    Example: to redact variables ending with ``_TOKEN`` or containing ``SECRET``::
+
+        set -g fish_redact_vars '*_TOKEN' '*_SECRET*'
```
I'd personally recommend against using environment variables to store secrets,
but I guess some legacy tools require it;
seems like a handy feature I guess.

If there are some commonly used variables out there (are there?),
we can maybe ship a reasonable default value (as a global variable, set for interactive sessions, see __fish_config_interactive).

fixing-things-enjoyer replied:
My use case is indeed screensharing. I've been juggling as many LLM CLI tools as I can get my hands on lately - [crush](https://github.com/charmbracelet/crush?tab=readme-ov-file#getting-started), [goose](https://block.github.io/goose/docs/guides/environment-variables/), [gemini-cli]{https://github.com/google-gemini/gemini-cli/blob/4828e4daf198a675ce118cec08dcfbd0bfbb28a6/docs/cli/configuration.md#environment-variables--env-files}, [openhands-cli](https://docs.all-hands.dev/usage/how-to/cli-mode#running-with-docker), [qwen-code](https://github.com/QwenLM/qwen-code/blob/93f5e59710ca192ef4ab739686cd26bc5898190d/docs/cli/configuration.md#environment-variables--env-files), [claude-code](https://docs.anthropic.com/en/docs/claude-code/settings#environment-variables)

Popularity ebbs and flows on these, most of them have mechanisms aside from environment variables to specify your LLM provider API Keys, but i find convenience in that they'll often use the same environment variable names, so like `GEMINI_API_KEY, OPENAI_API_KEY, OPENROUTER_API_KEY, etc...`. If I were choosing defaults, I wouldn't be mad at `*_API_KEY`, `*_TOKEN`, `*_SECRET`. 

That idea of a more comprehensive presentation mode sounds better than what I did here. Although I guess users could combine the existing defacto "private mode" and this redaction mechanism via their own custom functions.

krobelus commented on src/complete.rs:
```diff
@@ -1682,8 +1682,23 @@
+                    };
+
+                    if redacted {
+                        desc = wgettext!("Variable: [redacted]").to_owned();
```
we should be consistent with COMPLETE_VAR_DESC_VAL,
so either define COMPLETE_VAR_REDACTED_DESC_VAL,
or inline COMPLETE_VAR_DESC_VAL 
(if we do that, we should make sure that Lazy doesn't add any caching here,
and exists only because `wgettext!()` is not const.. I'm pretty sure that's the case)

fixing-things-enjoyer replied:
Went with existing pattern of defining a `COMPLETE_VAR_REDACTED_DESC_VAL`

krobelus commented on tests/pexpects/redact_vars.py:
```diff
@@ -0,0 +1,92 @@
+# Set TERM in the environment before spawning fish. This is more robust
+# and ensures the shell initializes with the correct terminal type.
+env = os.environ.copy()
+env["TERM"] = "xterm"
```
the comment sounds like an insecure LLM, maybe delete it.
The actual reason is that otherwise (with the default TERM=dumb), no completions will be displayed.

fixing-things-enjoyer replied:
got rid of it.

krobelus commented on tests/pexpects/redact_vars.py:
```diff
@@ -0,0 +1,92 @@
+
+# Reset key bindings to ensure a clean test environment.
+sendline("bind --erase --all")
+expect_prompt()
```
shouldn't need this, the environment is already clean

fixing-things-enjoyer replied:
got rid of it.

krobelus commented on tests/pexpects/redact_vars.py:
```diff
@@ -0,0 +1,92 @@
+
+# Set a low pager threshold to force it to appear for our list of completions.
+sendline("set -g fish_pager_completion_threshold 1")
+expect_prompt()
```
I don't think this variable exists

fixing-things-enjoyer replied:
got rid of it.

krobelus commented on tests/pexpects/redact_vars.py:
```diff
@@ -0,0 +1,92 @@
+expect_prompt()
+
+# Configure the redaction rules: a wildcard for `*_SECRET` and one explicit name.
+sendline("set -g fish_redact_vars '*_SECRET' MY_PUBLIC_VALUE")
```
why is MY_PUBLIC_VALUE chosen as name for something we want to hide from the public

fixing-things-enjoyer replied:
good question, swapped it to something more coherent with test:
`MY_EXPLICIT_SECRET` i.e. the non-wildcard based one we expect to be redacted.

krobelus commented on tests/pexpects/redact_vars.py:
```diff
@@ -0,0 +1,92 @@
+try:
+    sp.spawn.expect(pexpect.TIMEOUT, timeout=0.5)
+except pexpect.TIMEOUT:
+    pass
```
there's probably a better way to do this, buit I guess.
Would be nice if the comment was a bit less hand-wavy;
overall it would be fine to reduce comments a bit.

krobelus commented on tests/pexpects/redact_vars.py:
```diff
@@ -0,0 +1,92 @@
+assert "$MY_OTHER_VALUE (Variable: other_value)" in output_text
+assert "$MY_NORMAL_VAR (Variable: normal_value)" in output_text
+
+# Clean up after the test.
```
no need to clean up, each test gets its own PTY and fish process

fixing-things-enjoyer commented on doc_src/interactive.rst:
```diff
@@ -56,6 +56,15 @@ It also provides a large number of program specific scripted completions. Most o
+    If a variable name matches one of these patterns, its value will be replaced with ``[redacted]`` in the completion description.
+    This is useful for preventing accidental exposure of sensitive information like API keys or tokens during tab completion.
+
+    Example: to redact variables ending with ``_TOKEN`` or containing ``SECRET``::
```
Yep! updated.

fixing-things-enjoyer commented on src/complete.rs:
```diff
@@ -1682,8 +1682,23 @@ impl<'ctx> Completer<'ctx> {
+                        redact_patterns.as_list().iter().any(|pattern| {
+                            let wc = crate::parse_util::parse_util_unescape_wildcards(pattern);
+                            wildcard_match(&env_name, wc, false)
+                        })
```
refactored this accordingly and then went with clippy's suggestion.

krobelus commented:
> If I were choosing defaults, I wouldn't be mad at `*_API_KEY`,
> `*_TOKEN`, `*_SECRET`.

Sounds good to me, let's do that.

fixing-things-enjoyer commented:
> If I were choosing defaults, I wouldn't be mad at `*_API_KEY`, `*_TOKEN`, `*_SECRET`.
> Sounds good to me, let's do that.

added those defaults, adjusted docs to mention them, and tests to ensure they're present

krobelus commented on tests/pexpects/redact_vars.py:
```diff
@@ -0,0 +1,71 @@
+)
+
+# Verify expected redactions
+assert "$MY_API_KEY (Variable: [redacted])" in output_text
```
these assertions are not ideal because they don't print the actual fish output on failure.
But it's fine to ignore that for now.
